[benedictleejr] iP

benedictleejr opened the pull request:
# Wooper Pro Max
![Picture of the Pokemon, Wooper](https://assets.pokemon.com/assets/cms2/img/pokedex/full//194.png)
> “Woop Woop!” – Woopy the Wooper  

**Wooper Pro Max** is a simple yet powerful task manager designed to help you stay organized. It's:

- 📝 **Text-based**
- 🚀 **Easy to use**
- ⚡ **~~FAST~~ SUPER FAST**

## How to Get Started  

1. [Download the application](#).
2. Navigate to the downloaded directory, and type `java -jar ip.jar`.
3. Start adding your tasks!
4. Sit back and let **Wooper** manage them for you. :wink::+1:

## Features  

- :white_check_mark: Managing tasks  
- :hourglass_flowing_sand: Managing deadlines _(coming soon)_  
- :bell: Reminders _(coming soon)_  

## To-Do List  

- [x] Implement task addition  
- [x] Add task deletion  
- [ ] Implement reminders _(in progress)_  
- [ ] Add GUI _(coming soon)_  

## Code Snippet  

If you're a Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public static void main(String[] args) throws IOException {
    Wooper wooper = new Wooper();
    wooper.run();                
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

benedictleejr committed:
gitignore: Add .class files

benedictleejr committed:
Update Duke to Wooper
Rename, Greet, Exit

benedictleejr committed:
Add optimized IO methods

benedictleejr committed:
Add echo functionality
Add cosmetic changes for more personality

benedictleejr committed:
Add list functionality

benedictleejr committed:
Add Tasklist functionality

benedictleejr committed:
Add ToDos, Events, & Deadlines feature

benedictleejr committed:
Remove unnecessary imports

benedictleejr committed:
Add automated Text UI Testing functionality

benedictleejr committed:
Add JavaDocs comments for methods in ToDo, Deadline and Event

benedictleejr committed:
Add delete task functionality

benedictleejr committed:
Push all tags from week 2's work

benedictleejr committed:
Fix formatting inconsistencies

benedictleejr committed:
Add save functionality and persistent memory to tasklist

benedictleejr committed:
Merge branch 'branch-Level-7'

benedictleejr committed:
Add core dates and times functionality

benedictleejr committed:
Add command to print deadlines and events occurring on a specific date

benedictleejr committed:
Merge branch 'branch-Level-8'

benedictleejr committed:
Refactor into 3 new classes: Ui, Parser, Storage

Extract out all Ui functionality to Ui class, and storage functionality to storage class.
Parser class still WIP.

benedictleejr committed:
Refactor Wooper class onto Parser to handle parsing user commands

benedictleejr committed:
Merge branch 'branch-A4-Increments'

benedictleejr committed:
Put all classes into Wooper package

benedictleejr committed:
Merge remote-tracking branch 'origin/add-gradle-support'

benedictleejr committed:
Add JUnit Tests for Wooper and Tasklist classes

benedictleejr committed:
Add Gradle support for making Jar files

benedictleejr committed:
Add JavaDoc comments to almost all non-public classes/methods

benedictleejr committed:
Merge branch 'branch-A-JavaDoc'

benedictleejr committed:
Edit all Java files to adhere to coding standard

benedictleejr committed:
Merge branch 'branch-A-CodingStandard'

benedictleejr committed:
Add find tasks by keyword functionality

benedictleejr committed:
Merge branch 'branch-Level-9'

benedictleejr committed:
Fix typo in findTasks method

benedictleejr committed:
Add checkstyle and fix code to comply with coding convention

benedictleejr committed:
Fix style errors

benedictleejr committed:
Add JavaFX functionality - WIP

benedictleejr committed:
Refactor directory configurations

benedictleejr committed:
Implement fully functional GUI for Wooper

benedictleejr committed:
Fix code to adhere to coding conventions

bryonk commented on src/main/java/wooper/Parser.java:
```diff
@@ -0,0 +1,97 @@
+    /**
+     * Enum for the valid command types
+     */
+    public enum CommandType {
+        EXIT, LIST, TODO, DEADLINE, EVENT, MARK, UNMARK, DELETE, VIEW, INVALID, FIND
+    }
```
Good use of enum to limit the functions.

bryonk commented on src/main/java/wooper/Wooper.java:
```diff
@@ -0,0 +1,300 @@
+        String[] l = input.split(" ");
+        Parser.CommandType command = parser.parseCommand(input);
+
+        switch (command) {
+        case EXIT:
+            storage.saveTasks(FILE_PATH, tasks.getAllTasks());
+            response.append("Goodbye! See you next time.");
+            break;
+        case LIST:
+            response.append(handleList());
+            break;
+        case DELETE:
+            response.append(handleDelete(l));
+            break;
+        case VIEW:
+            response.append(handleView(l));
+            break;
+        case FIND:
+            response.append(handleFind(l));
+            break;
+        case MARK:
+        case UNMARK:
+            response.append(handleMarking(command, l));
+            break;
+        case TODO:
+            response.append(handleTodo(l));
+            break;
+        case DEADLINE:
+            response.append(handleDeadline(l));
+            break;
+        case EVENT:
+            response.append(handleEvent(l));
+            break;
+        default:
+            response.append("Invalid command.");
+            break;
+        }
```
Nice indentation on the switch statement.

bryonk commented on src/main/java/wooper/Wooper.java:
```diff
@@ -0,0 +1,300 @@
+     * @param l The string array containing the user input.
+     * @return Formatted String of tasks on specified date
+     */
+    public String handleView(String[] l) {
+        if (l.length < 2) { // Invalid input, prompt user
+            return "Please enter a date to view";
+        }
+        String date = l[1];
+        ArrayList<Task> tasksOnDate = tasks.getTasksOnDate(date);
+        StringBuilder sb = new StringBuilder();
+        for (int i = 0; i < tasksOnDate.size(); i++) {
+            Task t = tasksOnDate.get(i);
+            sb.append(String.format("%d.[%s][%s] %s\n", i + 1, t.getTaskType(), t.getStatusIcon(), t.getDescription()));
+        }
+        return sb.toString().trim();
+    }
```
Just a suggestion, could the handle[Function] functions be abstracted into a Command (class/interface) with [Function]Command children (i.e. AddCommand, ListCommand) to declutter the Wooper.java class, as some of the handle[Function] functions have the same input and output types and could be replaced by a single function.

bryonk reviewed (commented):
LGTM (nice pokemon picture in the description), overall coding standard is good and left a small suggestion on the code.

jaylontan commented on src/main/java/wooper/DialogBox.java:
```diff
@@ -0,0 +1,89 @@
+ */
+public class DialogBox extends HBox {
+    private static final String OPENING_MESSAGE = """
+            Hello! I'm Wooper, your personal tasklist chatbot.
+            How can I help you today?
+
+            Valid Commands:
+            1. Add tasks:
+                a. todo <task description>
+                b. deadline <task description> /by <due date> <due time>
+                c. event <task description> /from <start date> <start time> /to <end date> <end time>
+                NOTE: Date and time should be in the format YYYY-MM-DD HH:MM
+            2. View tasks:
+                a. list - views all tasks
+                b. view <date> - views tasks due on a specific date
+                c. find <keyword> - finds tasks with a specific keyword
+            3. Mark/unmark tasks as done:
+                a. mark <task number>
+                b. unmark <task number>
+            4. Delete tasks:
+                a. delete <task number>
+            5. Type 'exit' to exit
+            """;
+
```
Good description to notify users on how to properly use the chatbot with valid commands.

jaylontan commented on src/main/java/wooper/Parser.java:
```diff
@@ -0,0 +1,97 @@
+     * @param action User input.
+     * @return Returns the command type.
+     */
+    public CommandType parseCommand(String action) {
```
Can use switch-case statements rather than so many if blocks.

jaylontan commented on src/main/java/wooper/Wooper.java:
```diff
@@ -0,0 +1,300 @@
+        }
+    }
+
+}
```
Possible to abstract all these functions into TaskList.java

jaylontan commented on src/main/java/wooper/Storage.java:
```diff
@@ -0,0 +1,60 @@
+        try (ObjectInputStream ois = new ObjectInputStream(new FileInputStream(filePath))) {
+            tasklist = (ArrayList<Task>) ois.readObject();
+
+        } catch (FileNotFoundException e) { // if file not found, then create the new file
```
Handles error when file not found. Good!

benedictleejr committed:
# Add assert statements and JUnit tests

Assert statements ensure that important assumptions hold in the program
One JUnit test added for more test coverage

More assertions and JUnit tests to come, to improve overall test coverage

benedictleejr committed:
# Refactor code to improve code quality
As part of the move towards cleaner code.

## Refactors:
- Add JavaDocs for public methods
- Improved comment quality
- Use the default branch in if/else statement

benedictleejr committed:
# Add Streams to Tasklist class

To improve code conciseness, this commit rewrites the getTasksOnDate()
method in tasklist class to use streams

benedictleejr committed:
Merge pull request #1 from benedictleejr/branch-A-Assertions

# Add assert statements and JUnit tests

benedictleejr committed:
Merge branch 'master' into branch-A-CodeQuality

benedictleejr committed:
Merge pull request #2 from benedictleejr/branch-A-CodeQuality

# Refactor code to improve code quality

benedictleejr committed:
Merge branch 'master' into branch-A-Streams

benedictleejr committed:
Merge pull request #3 from benedictleejr/branch-A-Streams

# Add Streams to Tasklist class

benedictleejr committed:
# Add github actions workflow

Mainly targeting gradle commands like gradle clean

benedictleejr committed:
Add notebook and note class

benedictleejr committed:
Add notetaking functionality and major refactor

Created CommandHandler class to takeover all the logic from Wooper
class
Also allowed for flexible handling and reuse of code for tasks/notes

benedictleejr committed:
Merge pull request #4 from benedictleejr/branch-D-Notes

Branch d notes

benedictleejr committed:
Fix Javadoc spacing in Wooper class

benedictleejr committed:
Fix javadoc in ToDo class

benedictleejr committed:
Add User Guide

benedictleejr committed:
Fix image not showing up issue

benedictleejr committed:
Fix Javadoc format for all methods

benedictleejr committed:
Update README

benedictleejr committed:
Fix README image
